packages/flutter/lib/src/material/search.dart

```diff
@@ -0,0 +1,407 @@
+    widget.delegate._focusNode.addListener(_onFocusChanged);
+  }
+
+  @override
+  void dispose() {
+    super.dispose();
+    _controller.removeListener(_onQueryChanged);
+    widget.animation.removeStatusListener(_onAnimationStatusChanged);
+    widget.delegate._currentBodyNotifier.removeListener(_onSearchBodyChanged);
+    widget.delegate._focusNode.removeListener(_onFocusChanged);
+  }
+
+  void _onAnimationStatusChanged(AnimationStatus status) {
+    if (status != AnimationStatus.completed) {
+      return;
+    }
+    widget.animation.removeStatusListener(_onAnimationStatusChanged);
+    if (widget.delegate._currentBody == _SearchBody.suggestions) {
+      FocusScope.of(context).requestFocus(widget.delegate._focusNode);
+    }
+  }
+
+  void _onFocusChanged() {
+    if (widget.delegate._focusNode.hasFocus && widget.delegate._currentBody != _SearchBody.suggestions) {
+      widget.delegate.showSuggestions(context);
+    }
+  }
+
+  void _onQueryChanged() {
+    setState(() {
+      // rebuild ourselves because query changed.
+    });
+  }
+
+  void _onSearchBodyChanged() {
+    setState(() {
+      // rebuild ourselves because search body changed.
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final ThemeData theme = widget.delegate.appBarTheme(context);
+    Widget body;
+    switch(widget.delegate._currentBody) {
+      case _SearchBody.suggestions:
+        body = new Container(
+          key: const ValueKey<_SearchBody>(_SearchBody.suggestions),
+          child: widget.delegate.buildSuggestions(context),
+        );
+        break;
+      case _SearchBody.results:
+        body = new Container(
+          key: const ValueKey<_SearchBody>(_SearchBody.results),
+          child: widget.delegate.buildResults(context),
+        );
+        break;
+    }
+
+
+    return new Scaffold(
+      appBar: new AppBar(
+        backgroundColor: theme.primaryColor,
+        iconTheme: theme.primaryIconTheme,
+        textTheme: theme.primaryTextTheme,
+        brightness: theme.primaryColorBrightness,
+        leading: widget.delegate.buildLeading(context),
+        // TODO(goderbauer): Show the search key (instead of enter) on keyboard, https://github.com/flutter/flutter/issues/17525
+        title: new TextField(
+          controller: _controller,
+          focusNode: widget.delegate._focusNode,
+          style: theme.textTheme.title,
+          onSubmitted: (String _) {
+            widget.delegate.showResults(context);
+          },
+          decoration: new InputDecoration(
+            border: InputBorder.none,
+            hintText: MaterialLocalizations.of(context).searchFieldLabel
+          ),
+        ),
+        actions: widget.delegate.buildActions(context),
+      ),
+      body: new AnimatedSwitcher(
+        duration: const Duration(milliseconds: 300),
+        child: body,
+      ),
+    );
+  }
+
+  TextEditingController get _controller => widget.delegate._textEditingController;
```

Comment:
It might be a little clearer if this was called queryTextController or something less generic.